lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskHostBase.cs

```diff
@@ -0,0 +1,229 @@
+using Org.Apache.REEF.IMRU.OnREEF.Driver;
+using Org.Apache.REEF.Network.Group.Task;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Remote.Impl;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
+{
+    internal abstract class TaskHostBase
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(TaskHostBase));
+
+        /// <summary>
+        /// Specify whether to invoke garbage collector or not
+        /// </summary>
+        protected readonly bool _invokeGc;
+
+        /// <summary>
+        /// Shows if the object has been disposed.
+        /// </summary>
+        private int _disposed;
+
+        /// <summary>
+        /// Group Communication client for the task
+        /// </summary>
+        private readonly IGroupCommClient _groupCommunicationsClient;
+
+        /// <summary>
+        /// CommunicationGroupClient for the task
+        /// </summary>
+        protected readonly ICommunicationGroupClient _communicationGroupClient;
+
+        /// <summary>
+        /// Task close Coordinator to handle the work when receiving task close event
+        /// </summary>
+        private readonly TaskCloseCoordinator _taskCloseCoordinator;
+
+        /// <summary>
+        /// The cancellation token to control the group communication operation cancellation
+        /// </summary>
+        protected readonly CancellationTokenSource _cancellationSource;
+
+        /// <summary>
+        /// Task host base class to hold the common stuff of both map and update tasks
+        /// </summary>
+        /// <param name="groupCommunicationsClient">Group Communication Client</param>
+        /// <param name="taskCloseCoordinator">The class that handles the close event for the task</param>
+        /// <param name="invokeGc">specify if want to invoke garbage collector or not </param>
+        protected TaskHostBase(
+            IGroupCommClient groupCommunicationsClient,
+            TaskCloseCoordinator taskCloseCoordinator,
+            bool invokeGc)
+        {
+            _groupCommunicationsClient = groupCommunicationsClient;
+            _communicationGroupClient = groupCommunicationsClient.GetCommunicationGroup(IMRUConstants.CommunicationGroupName);
+
+            _invokeGc = invokeGc;
+            _taskCloseCoordinator = taskCloseCoordinator;
+            _cancellationSource = new CancellationTokenSource();
+        }
+
+        /// <summary>
+        /// Handle the exceptions in the Call() method
+        /// Default to IMRUSystemException to make it recoverable
+        /// </summary>
+        /// <param name="memento"></param>
+        /// <returns></returns>
+        public byte[] Call(byte[] memento)
+        {
+            Logger.Log(Level.Info, "Entering {0} Call().", TaskHostName);
+            try
+            {
+                return TaskBody(memento);
+            }
+            catch (Exception e) 
+            {
+                if (IsCommunicationException(e))
```

Comment:
We need to consider one more case here: if an exception happened in `_mapTask.Map` or `_updateTask.Update` or another part of client code, we catch it, wrap it in `IMRUTaskAppException` and re-throw it, and then here we catch this `IMRUTaskAppException` and wrap it in `IMRUSystemException` because it's not communication exception (and we never produce `IMRUTaskAppException`). If we get `IMRUTaskAppException` we should throw it as is.

*We should also update `TestFailMapperTasks` which now incorrectly assumes that in case of app exception we should retry.